compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -11250,20 +11250,29 @@ private STNode parseTemplateContentAsRegExp() {
         // Separate out the interpolated expressions to a queue. Then merge the string content using '${}'.
         // These '${}' are used to represent the interpolated locations. Regular expression parser will replace '${}'
         // with the actual interpolated expression, while building the regular expression tree.
+        this.tokenReader.startMode(ParserMode.REGEXP);
         ArrayDeque<STNode> expressions = new ArrayDeque<>();
         StringBuilder regExpStringBuilder = new StringBuilder();
         STToken nextToken = peek();
+        boolean visitedCharacterClassStartToken = nextToken.text().equals("[");
```

Comment:
I think we don't need to start a new lexer mode and capture `[` and `]` as separate tokens.

Can't we simply do, `boolean visitedCharacterClassStartToken = LexerTerminals.OPEN_BRACKET == nextToken.text().charAt(0)` here? 
Since we anyway contacting `[` and `]` here, tokenizing it at BallerinaLexer.java is unnecessary.

Reply:
It does not always start with [ and end with ].
eg:
```ballerina
    _ = re `([^a-b${b-a}])`;
```

Reply:
Also, Let's add a small comment with the example `[${abc}]`.

Reply:
Oh right. my bad. How about at the ballerina lexer we keep a flag for `[`. If we reach it we do not allow interpolation tokenizing logic.
https://github.com/ballerina-platform/ballerina-lang/blob/2bee70a2e107474399ce8ef9601bf8712849b6a5/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaLexer.java#L1472-L1473

Here we add the flag inside the if condition. Then we don't need the concatenation part in the parser.

Reply:
Lexer sometimes tokenizes multiple characters into single literal unless we handle it specifically in https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaLexer.java#L1468
If we are to special case that and return just [ or ] when it sees them, we need a way to differentiate regexps. If not we'll be doing it for all templates. I think the parser contexts would be a wiser choice since it will be anyway a hash lookup.

Reply:
I mean we can avoid tokenizing and concatinating them back. For example `([${abc}])` will be tokenized as a single string template token at the ballerina lexer. We will have the `ParserMode.REGEXP`. (to distinguish from normal template expr).
Also, instead of `this.mode == ParserMode.REGEXP` checks at several places, we can pass down a flag like this,
```java
        switch (this.mode) {
            case TEMPLATE:
                token = readTemplateToken(false);
                break;
            case REGEXP:
                token = readTemplateToken(true); // boolean isRegExp = true;
                break;
```